src/tests/math/test_linspace.f90

```diff
@@ -0,0 +1,396 @@
+program test_linspace
+    use stdlib_error, only: check
+    use stdlib_kinds, only: sp, dp, int16, int8
+    use stdlib_math, only: linspace, DEFAULT_LINSPACE_LENGTH
+
+    implicit none
+
+    integer :: iunit
+    logical :: warn = .false.
+    real(dp), parameter :: TOLERANCE = 0.00001 ! Percentage of the range for which the actual gap must not exceed
+
+    ! Testing linspace.
+    !
+    ! For single and double precision, check if the beginning and end values are properly recorded
+    ! and make sure that the size of the result array is as expected.
+    !
+    ! This testing suite makes use of the a repeated section of code that will check to make
+    ! sure that every element is linearly spaced (i.e., call check(|array(i+1) - array(i)| < |expected_value| * TOLERANCE)).
+    ! I would convert this repeated code into a subroutine but that would require the implementation of a
+    ! generic procedure given that each linear space will have a different expected_value type and kind.
+
+    open(newunit=iunit, file="test_linspace_log.txt", status="unknown") ! Log the results of the functions
+
+    write(iunit,*) "Writing to unit #: ", iunit
+
+    call test_linspace_sp
+    call test_linspace_dp
+    call test_linspace_neg_index ! Make sure that when passed a negative index the result is an empty array
+    call test_linspace_cmplx
+    call test_linspace_cmplx_2
+    call test_linspace_cmplx_3
+    call test_linspace_cmplx_sp
+    call test_linspace_cmplx_sp_2
+    call test_linspace_int16
+    call test_linspace_int8
+
+    close(unit=iunit)
+
+contains
+
+    subroutine test_linspace_sp
+
+        integer, parameter :: n = 20
+        real(sp) :: start = 1.0_sp
+        real(sp) :: end = 10.0_sp
+        real(sp) :: expected_interval
+        real(sp) :: true_difference
+
+        integer :: i
+        real(sp), allocatable :: x(:)
+
+        x = linspace(start, end, n)
+
+        expected_interval =( end - start ) / real(( n - 1 ), sp)
+
+        call check(x(1) == start, msg="Initial value of array is not equal to the passed start parameter", warn=warn)
+        call check(x(n) == end, msg="Final array value is not equal to end parameter", warn=warn)
+        call check(size(x) == n, msg="Array not allocated to appropriate size", warn=warn)
+
+        print *, "Made it through first round of tests"
+
+        ! Due to roundoff error, it is possible that the jump from x(n-1) to x(n) is slightly different than the expected interval
+        do i = 1, n-1
+
+            true_difference = x(i + 1) - x(i)
+            call check(abs(true_difference - expected_interval) < abs(expected_interval) * TOLERANCE)
```

Comment:
Why not:

```suggestion
            call check(abs(true_difference - expected_interval) <  TOLERANCE)
```

Reply:
I originally only compared against a fixed tolerance. However, given the very nature of **floating** (and not fixed) point numbers, it makes more sense to compare against an EPSILON value that scales with the magnitude of the numbers in question.

i.e, we should allow a higher **magnitude** epsilon when testing the interval between
```fortran
linspace(1.0_dp, 1D100)
```
compared to 
```fortran
linspace(1.0_dp, 1.1_dp)
```

Reply:
Ok. It is why a constant is added in the definition of e.g., `TOLERANCE_DP` above.